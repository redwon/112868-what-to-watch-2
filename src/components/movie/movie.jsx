import React, {Fragment} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import {MovieType} from '../../types';
import {getRelatedMovies, getMovieById} from '../../selectors';
import {ActionCreator} from '../../reducer/genre/genre';

import Header from '../header/header';
import Footer from '../footer/footer';
import MoviesList from '../movies-list/movies-list';
import MovieTabs from '../movie-tabs/movie-tabs';

import withActiveItem from '../../hocs/with-active-item/with-active-item';

const MovieTabsWrapped = withActiveItem(MovieTabs);

const Movie = ({movie, relatedMovies}) => {
  if (!movie) {
    return null;
  }

  return (
    <Fragment>
      <section
        className="movie-card movie-card--full"
        style={{backgroundColor: movie.backgroundColor}}
      >
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={movie.backgroundImage} alt={movie.name} />
          </div>

          <Header />

          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{movie.name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{movie.genre}</span>
                <span className="movie-card__year">{movie.released}</span>
              </p>

              <div className="movie-card__buttons">
                <button className="btn btn--play movie-card__button" type="button">
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </button>
                <button className="btn btn--list movie-card__button" type="button">
                  <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"></use>
                  </svg>
                  <span>My list</span>
                </button>
                <a href="add-review.html" className="btn movie-card__button">
                  Add review
                </a>
              </div>
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img
                src={movie.posterImage}
                alt={movie.name}
                width="218"
                height="327"
              />
            </div>

            <div className="movie-card__desc">
              <MovieTabsWrapped movie={movie} />
            </div>
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>

          <MoviesList movies={relatedMovies} />
        </section>

        <Footer />
      </div>
    </Fragment>
  );
};

Movie.propTypes = {
  movie: MovieType,
  relatedMovies: PropTypes.arrayOf(MovieType),
  match: PropTypes.object,
};

const mapStateToProps = (state, props) => Object.assign({}, props, {
  movie: getMovieById(state, props.match.params.id),
  relatedMovies: getRelatedMovies(state, props.match.params.id),
});

const mapDispatchToProps = (dispatch) => ({
  onGenreChange: (genre) => {
    dispatch(ActionCreator.changeGenre(genre));
  },
});

export {Movie};

export default connect(mapStateToProps, mapDispatchToProps)(Movie);
