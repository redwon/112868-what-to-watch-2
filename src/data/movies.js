const movies = [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    img: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`
  },
  {
    name: `Bohemian Rhapsody`,
    img: `img/bohemian-rhapsody.jpg`
  },
  {
    name: `Macbeth`,
    img: `img/macbeth.jpg`
  },
  {
    name: `Mindhunter`,
    img: `img/mindhunter.jpg`
  }
];

export default movies;
