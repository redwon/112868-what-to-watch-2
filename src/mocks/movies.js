const movies = [
  {
    id: 1,
    name: `War of the Worlds`,
    posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg`,
    previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg`,
    backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg`,
    backgroundColor: `#9B7E61`,
    description: `Lorem ipsum`,
    rating: 9.3,
    scoresCount: 386834,
    director: `Steven Spielberg`,
    starring: [`Tom Cruise`, `Dakota Fanning`, `Tim Robbins`],
    runTime: 116,
    genre: `Adventure`,
    released: 2005,
    isFavorite: false,
    videoLink: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  },
  {
    id: 2,
    name: `Bronson`,
    posterImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/bronson.jpg`,
    previewImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/bronson.jpg`,
    backgroundImage: `https://htmlacademy-react-2.appspot.com/wtw/static/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    description: `Lorem ipsum`,
    rating: 3.6,
    scoresCount: 109661,
    director: `Nicolas Winding Refn`,
    starring: [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    runTime: 92,
    genre: `Action`,
    released: 2008,
    isFavorite: false,
    videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
  }
];

export default movies;
