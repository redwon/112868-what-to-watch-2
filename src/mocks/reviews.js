const reviews = [
  {
    id: 1,
    user: {id: 1, name: `name`},
    rating: 3.7,
    comment: `comment 1`,
    date: `2019-11-29T14:11:47.482Z`
  },
  {
    id: 2,
    user: {id: 2, name: `name`},
    rating: 4.2,
    comment: `comment 2`,
    date: `2019-11-23T14:11:47.482Z`
  }
];

export default reviews;
