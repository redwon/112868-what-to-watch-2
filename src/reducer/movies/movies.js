export const initialState = [];

export const ActionType = {
  LOAD_MOVIES: `LOAD_MOVIES`,
};

export const ActionCreator = {
  loadMovies: (movies) => ({
    type: ActionType.LOAD_MOVIES,
    payload: movies
  }),
};

export const Operations = {
  loadMovies: () => (dispatch, _, api) => {
    return api.get(`/films`)
      .then((response) => {
        dispatch(ActionCreator.loadMovies(response.data));
      });
  }
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_MOVIES:
      return action.payload;
  }

  return state;
};
