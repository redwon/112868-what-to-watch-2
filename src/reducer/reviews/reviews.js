export const initialState = [];

export const ActionType = {
  LOAD_REVIEWS: `LOAD_REVIEWS`,
};

export const ActionCreator = {
  loadReviews: (reviews) => ({
    type: ActionType.LOAD_REVIEWS,
    payload: reviews
  })
};

export const Operations = {
  loadReviews: (movieId) => (dispatch, _, api) => {
    return api.get(`/comments/${movieId}`)
      .then((response) => {
        dispatch(ActionCreator.loadReviews(response.data));
      });
  }
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_REVIEWS:
      return action.payload;
  }

  return state;
};
