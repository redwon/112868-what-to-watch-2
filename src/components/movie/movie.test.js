import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';

import movies from '../../mocks/movies';
import reviews from '../../mocks/reviews';

import {Movie} from './movie';

jest.mock(`../header/header`, () => () => `Header`);
jest.mock(`../add-my-list/add-my-list`, () => () => `AddMyList`);

it(`renders correctly`, () => {
  const tree = renderer
    .create(<BrowserRouter>
      <Movie
        movie={movies[0]}
        reviews={reviews}
        relatedMovies={movies.slice(0, 4)}
        onGenreChange={jest.fn()}
      />
    </BrowserRouter>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
