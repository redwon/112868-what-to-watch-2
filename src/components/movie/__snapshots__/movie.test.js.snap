// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    style={
      Object {
        "backgroundColor": "#9B7E61",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="War of the Worlds"
          src="https://htmlacademy-react-2.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg"
        />
      </div>
      Header
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            War of the Worlds
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Adventure
            </span>
            <span
              className="movie-card__year"
            >
              2005
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              onClick={[Function]}
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="add-review.html"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="War of the Worlds"
            height="327"
            src="https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item movie-nav__item--active"
              >
                <a
                  className="movie-nav__link"
                  href="#"
                  onClick={[Function]}
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item "
              >
                <a
                  className="movie-nav__link"
                  href="#"
                  onClick={[Function]}
                >
                  Details
                </a>
              </li>
              <li
                className="movie-nav__item "
              >
                <a
                  className="movie-nav__link"
                  href="#"
                  onClick={[Function]}
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div
            className="movie-rating"
          >
            <div
              className="movie-rating__score"
            >
              9.3
            </div>
            <p
              className="movie-rating__meta"
            >
              <span
                className="movie-rating__level"
              >
                Very good
              </span>
              <span
                className="movie-rating__count"
              >
                386834
                 ratings
              </span>
            </p>
          </div>
          <div
            className="movie-card__text"
          >
            Lorem ipsum
            <p
              className="movie-card__director"
            >
              <strong>
                Director: 
                Steven Spielberg
              </strong>
            </p>
            <p
              className="movie-card__starring"
            >
              <strong>
                Starring: 
                Tom Cruise
                Dakota Fanning
                Tim Robbins
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              className="player__video"
              muted={true}
              poster="https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg"
              src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="#"
              onClick={[Function]}
            >
              War of the Worlds
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              className="player__video"
              muted={true}
              poster="https://htmlacademy-react-2.appspot.com/wtw/static/film/poster/bronson.jpg"
              src="https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4"
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="#"
              onClick={[Function]}
            >
              Bronson
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 
          2019
           What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
